### ISSUE TITLE
C# client read operation hangs if there are network problems

### ISSUE THREAD

**Issue by gurkerlbaron** [RicoSuter/NSwag#4953]
I'm using code generator v14.1.0.0 to generate C# client code. My code runs on .net core 8.0.303.
In my project I had several issues with hanging client request, they never finished. I found out that this behavior was related to network overload (4Mbit link) and dropped TCP/IP packets. This leaded to a hanging call of:

```
protected virtual async System.Threading.Tasks.Task<ObjectResponseResult<T>> ReadObjectResponseAsync<T>(System.Net.Http.HttpResponseMessage response, System.Collections.Generic.IReadOnlyDictionary<string, System.Collections.Generic.IEnumerable<string>> headers, System.Threading.CancellationToken cancellationToken)
```

The underlying HttpClient timeout does not work on the stream. The TCP failure is not detected, because of TCP keep alive not enabled on HttpClient. The problematic code seems to be:

```
try
{
    using (var responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
    using (var streamReader = new System.IO.StreamReader(responseStream))
    using (var jsonTextReader = new Newtonsoft.Json.JsonTextReader(streamReader))
    {
        var serializer = Newtonsoft.Json.JsonSerializer.Create(JsonSerializerSettings);
        var typedBody = serializer.Deserialize<T>(jsonTextReader);
        return new ObjectResponseResult<T>(typedBody, string.Empty);
    }
}
catch (Newtonsoft.Json.JsonException exception)
{
    var message = "Could not deserialize the response body stream as " + typeof(T).FullName + ".";
    throw new ApiException(message, (int)response.StatusCode, string.Empty, headers, exception);
}
```

If the network link get lost during the rx data transfer, the responseStream never timed out and the method hangs forever.
I tried to add a cancelationToken with timeout to ReadAsStreamAsync but it event did not help.
The working solution was to add a timeouthandling and dispose the underlying stream:

```
try
  {
      _log.Info($"ReadAsStreamAsync");
      var tc = new CancellationTokenSource(60000);      <--- used for timeout
      using (var responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
      using (tc.Token.Register(() => { _log.Info($"responseStream.Dispose()"); responseStream.Dispose();}))
      using (var streamReader = new System.IO.StreamReader(responseStream))
      using (var jsonTextReader = new Newtonsoft.Json.JsonTextReader(streamReader))
      {
          var serializer = Newtonsoft.Json.JsonSerializer.Create(JsonSerializerSettings);
          var typedBody = serializer.Deserialize<T>(jsonTextReader);
          _log.Info($"ReadObjectResponseAsync return");
          return new ObjectResponseResult<T>(typedBody, string.Empty);
      }
  }
  catch (Newtonsoft.Json.JsonException exception)
  {
      var message = "Could not deserialize the response body stream as " + typeof(T).FullName + ".";
      throw new ApiException(message, (int)response.StatusCode, string.Empty, headers, exception);
  }
```

It's not easy to reproduce the wrong behavior because you must drop network packets. But I tested the timeout in a different way. I made a request with large response (leads to a long transfer time on network) and I used a short timeout of about 10ms. The expected result was, that the call fails because of the timeout and that worked properly.
The important thing for testing is to abort the TCP connection in a way, that the operation system itself cannot detect it. (e.g. link down)
Because the problem occures during a download, the socket cannot detect the broken link without tcp keepalive enabled (on the server side it will be detected because of the missing tcp ack). So the TCP connection on client side is still active and on the server side dropped. Without any read timeout it's not possible to detect that situation and close the tcp connection on client side as well.

Perhaps there is a better solution, but I could not find one.

A similar problem can be found here: [https://github.com/dotnet/runtime/issues/36822](url)

**Comment by arshrp** [comment]
#5186 might fix this?

**Comment by ErikPilsits-RJW** [comment]
@arshrp I don't believe this will fix the problem when using a stream. `ReadAsStreamAsync` returns a stream immediately which is used by the json serializer to deserialize the response. It is here that cancellation must occur, however the token is already passed to the serializer. The real problem here is that the token is _not cancelled._

I've also run into this problem. Our solution was to create a linked token with an explicit 100 second timeout using `CancelAfter` and pass that token to the client methods. This is quite cumbersome and requires a change to every call site.

I think a better solution would be to provide access to the cancellation token or better yet a cancellation token source in a partial method somewhere. The source would be better because a `CancellationTokenSource` must be disposed. If we only get access to modify the token in a `PrepareRequest` method, and create our own linked token, we can't get access to the correct scope to manage the source's disposal.

I can't think of a really clean way to implement this however... a partial method that gets a linked token source created from the passed in token, and the new token gets passed to all the following method calls? Maybe just a client level property where the user can set an explicit timeout and that's handled internally?

cc: @RicoSuter

**Comment by arshrp** [comment]
I'm confused.  The methods on the SDK generated by NSwag have an overload that takes a CancellationToken, so the consumer of the client is in control of the token (as well as the source that generated it).  After my change, that token is now passed to the ReadAsStringAsync method used internally.

If your token isn't timing out then I'm not sure that's something NSwag can handle.  The token should be timing out if created correctly.

I tested this by generating a blocking HttpClient (causing ReadAsStringAsync to never return) and passing that to the NSwag client.  When the token timed out, the NSwag method properly threw a CancellationException.

Can you retry your issue using the latest master branch, which has my change?  If it truly doesn't fix it then I don't think I properly understand your issue.

**Comment by ErikPilsits-RJW** [comment]
It's not a bug in nswag.. it's a missing solution for this scenario. Tokens don't cancel on their own. So in order to accomplish this I need to create a new linked token source and set it up to cancel before every client api call. It's just cumbersome. It would be more convenient if it was handled inside the generated client.

Just a point of clarification, you mentioned `ReadAsStringAsync`, but mostly `ReadAsStreamAsync` is used. That flow never had the issue your PR addresses, because the token was already passed to the serializer. But that doesn't matter, it's not the real issue.

I'm going to try out a change to the liquid template, and maybe that will help you understand what I'm getting at.

**Comment by arshrp** [comment]
I don't think most users would want NSwag to generate tokens, they would want control over them.  At my end, I create tokens that expire on a set time frame, ex: new CancellationTokenSource(TimeSpan.FromSeconds(10)), those will in fact cancel on their own once the allotted time expires, causing the NSwag method to throw an exception, which works perfect for me.

The call to ReadAsStreamAsync did have the issue my PR fixes.  Have a look [here](https://github.com/RicoSuter/NSwag/pull/5189/files#diff-ed4ff5dc274259830c56196e45f9fb86e0dbfd4e8678145945f6645aa17185dc)

Are you able to recreate your issue using the master version of NSwag, which has my merged change?  You can use this package reference from https://www.myget.org/F/nswag/api/v3/index.json to test:

	<PackageReference Include="NSwag.ApiDescription.Client" Version="14.4.1-preview-20250616-1934">
		<PrivateAssets>all</PrivateAssets>
		<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
	</PackageReference>

If my PR truly doesn't address your issue, then I will bow out with apologies :)

**Comment by ErikPilsits-RJW** [comment]
Your PR is fine. My point is that to provide this timeout, I have to create a linked token source and add the timeout _every time I make a client api call_. It's a lot of duplicated code. It would be a convenience if the nswag client provided a mechanism to centralize it.

Here's an initial idea:

![Image](https://github.com/user-attachments/assets/06bd5b81-cf2f-4f89-9073-a8e63bf36f66)

![Image](https://github.com/user-attachments/assets/a8e141c4-7c45-417a-8561-36f77e779bb7)

![Image](https://github.com/user-attachments/assets/5be8efea-6617-434f-8e99-13bd0c6f6997)

![Image](https://github.com/user-attachments/assets/be08e662-0501-49cf-adb1-7f074e3e03c7)

Then in the partial method, something simple like:

![Image](https://github.com/user-attachments/assets/95e01020-49f5-4863-b8d5-dcd0e7b9ced7)

This uses a familiar partial method pattern, is non-breaking, is optional, is transparent to the client, and provides a central place to configurate a client timeout. It would be a pretty easy change to the liquid template, which is what I'm going to try next.
